src/3d/terrain/quantizedmeshgeometry.cpp

```diff
@@ -0,0 +1,367 @@
+  quint16 *srcIdxPtr = ( quint16 * )dataPtr;
+  int highest = 0;
+  for ( uint i = 0; i < triangleCount * 3; ++i )
+  {
+    quint16 code = *srcIdxPtr++;
+    *indicesPtr++ = highest - code;
+    if ( code == 0 )
+      ++highest;
+  }
+
+  // TODO: edge indices
+
+  // TODO: extensions
+
+  //qDebug() << hdr.CenterX << " " << hdr.CenterY << " " << hdr.CenterZ;
+  //qDebug() << "VC " << vertexCount;
+  //qDebug() << "TC " << triangleCount;
+
+  return t;
+}
+
+
+#include <QGuiApplication>
+#include <QNetworkRequest>
+#include <QNetworkReply>
+#include "qgsnetworkaccessmanager.h"
+
+void QuantizedMeshGeometry::downloadTileIfMissing( int tx, int ty, int tz )
+{
+  // i am not proud of this bit of code... quick&dirty!
+
+  QString tileFilename = _tileFilename( tx, ty, tz );
+  if ( !QFile::exists( tileFilename ) )
+  {
+    qDebug() << "downloading tile " << tx << " " << ty << " " << tz;
+    bool downloaded = false;
+    QString url = QString( "http://assets.agi.com/stk-terrain/tilesets/world/tiles/%1/%2/%3.terrain" ).arg( tz ).arg( tx ).arg( ty );
+    QNetworkRequest request( url );
+    request.setRawHeader( QByteArray( "Accept-Encoding" ), QByteArray( "gzip" ) );
+    request.setRawHeader( QByteArray( "Accept" ), QByteArray( "application/vnd.quantized-mesh,application/octet-stream;q=0.9" ) );
+    request.setRawHeader( QByteArray( "User-Agent" ), QByteArray( "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/58.0.3029.110 Chrome/58.0.3029.110 Safari/537.36" ) );
+    QNetworkReply *reply = QgsNetworkAccessManager::instance()->get( request );
+    connect( reply, &QNetworkReply::finished, [reply, tileFilename, &downloaded]
+    {
+      QFile fOut( tileFilename );
+      fOut.open( QIODevice::WriteOnly );
+      fOut.write( reply->readAll() );
+      fOut.close();
+      reply->deleteLater();
+      downloaded = true;
+    } );
+
+    while ( !downloaded )
+    {
+      qApp->processEvents();
+    }
+  }
+}
+
+// --------------
+
+#include "qgscoordinatetransform.h"
+#include "qgsmaptopixel.h"
+#include "map3d.h"
+
+QuantizedMeshGeometry::QuantizedMeshGeometry( QuantizedMeshTile *t, const Map3D &map, const QgsMapToPixel &mapToPixel, const QgsCoordinateTransform &terrainToMap, QNode *parent )
+  : QGeometry( parent )
+{
+  int vertexCount = t->uvh.count() / 3;
+  int indexCount = t->indices.count();
+
+  int vertexEntrySize = sizeof( float ) * ( 3 + 2 );
+
+  double xMinWgs = t->extent.xMinimum();
+  double yMinWgs = t->extent.yMinimum();
+  double widthWgs = t->extent.width();
+  double heightWgs = t->extent.height();
+  QgsPointXY ptMinProjected = QgsPointXY( map.originX, map.originY );
+
+  QByteArray vb;
+  const qint16 *uvh = t->uvh.constData();
+  vb.resize( vertexCount * vertexEntrySize );
+  float *vbptr = ( float * ) vb.data();
+  for ( int i = 0; i < vertexCount; ++i )
+  {
+    qint16 u = uvh[i], v = uvh[vertexCount + i], h = uvh[vertexCount * 2 + i];
+    float uNorm = u / 32767.f;  // 0...1
```

Comment:
Do you think that `std::numeric_limits<int>::max()` could be used here?

Reply:
I do not have a strong preference but this seems easier to read to me - just a personal preference :)

(quantized mesh terrain generator is not active right now anyway)